packages/react-components/react-utilities/src/compose/types.ts

```diff
@@ -233,3 +234,29 @@ export type ForwardRefComponent<Props> = ObscureEventName extends keyof Props
 export type SlotClassNames<Slots> = {
   [SlotName in keyof Slots]-?: string;
 };
+
+/**
+ * A definition of a slot, as a component, very similar to how a React component is declared,
+ * but with some additional metadata that is used to determine how to render the slot.
+ */
+export type SlotComponent<Props extends UnknownSlotProps> = Props & {
```

Comment:
I've been considering some cleanup and updates to the `Slot` type, including creating a type called `SlotComponent` for slots of type component.

I'm not 100% sure I'll end up doing that, but it would be great to reserve the `SlotComponent` name for that use if possible. Also, this object isn't really a _component_, but more of a fancy props object. Could we call it something different? Options I can think of:
- `SlotState`
- `SlotObject`
- `ResolvedSlot`

Reply:
> this object isn't really a component, but more of a fancy props object

@behowell  what is a component but a fancy object with a function signature?!  this name has been inspired from the `ExoticComponent` signature, and as far as I understand a component is any object that can be interpreted by the pragma. In other words: `React.CreateElement` first argument is either a string or a component (which can be a function of a fancy object with a function signature), so the way I see `SlotComponent` is the perfect name for this case.

I believe you're talking of creating a type that will unify slot declaration, and a slot declaration on the PoV of the interface if not a component, but a shorthand of possible types. I'm not sure `SlotComponent` would be a proper name for that...

Reply:
I don't think these slot objects are really like components. A component is like `Button` or `Dialog`: a re-usable type that can be passed props to create elements. These objects are closer to the `ReactElement` type, which contains type info and props. But I don't think we should call them `SlotElement` either, since they also are not ReactElements: they're the building blocks necessary to create a ReactElement.

I like the idea of calling them `SlotState`, since they go with the `ComponentState` type: they contain mutable state about how to render the slot.

As for my potential use of `SlotComponent`, the idea was to split out the current `Slot` type into different types: SlotElement for slots of intrinsic element types (div, etc.), and SlotComponent for slots of component types, and reduce the reliance on type inference in the existing Slot type. It could potentially be named e.g. `Slot.Component` as well, which wouldn't technically conflict with this name, but would be awfully confusing.

I'm not entirely sure whether I'll use this name in the future, but it would be great not to use it for this case if possible.

Reply:
They're exactly equal to `Button` and `Dialog` in the sense that they will be the argument passed to the pragma method, right?! They're definitely closer to `ExoticComponent` type than `ReactElementType` (I literally copy pasted `ExoticComponent` on the first implementation), but I can see your point:

> In the user perspective the `SlotComponent` (or `Slot.Component`) is one of the options you can pass as the shorthand value on the component properties, and everything internally becomes just state.

My problem with `SlotState` is that you don't render state, you render elements. And a component is an element that is an object... So for the PoV of the library a `slot` is basically the same thing as a `Button` or a `Dialog` with this new pragma implementation (as it is something you render like `<slot/>`, similar to `<Button/>`).

The way I see what you're calling `Slot.Component` is also valid from the consumer perspective: it's a slot that is built from another component, something like `Slot.FromComponentType`?! but I agree that this would be too verbose.

It also seems like this "consumer" vs "library" PoV is also a problem on react types implementation, I guess that' why `React.ComponentType` existis?! so how about `SlotComponentType`? I'd be ok with it.

Reply:
How about `SlotObject`?

I suppose this is more of a philosophical question at this point: is this thing actually a component? I still think it isn't, because it _contains_ `SLOT_ELEMENT_TYPE` (the actual component it will use to render), as well as the props that will render with that component. It's also single-use and created each render, unlike a component like `Button`, which is static and doesn't change between renders. Granted, it's used _like_ a component with the custom JSX pragma and createElement function. 🤷‍♂️